src/java.base/share/classes/jdk/internal/constant/MethodTypeDescImpl.java

```diff
@@ -151,14 +167,17 @@ public MethodTypeDesc dropParameterTypes(int start, int end) {
         if (end < argTypes.length) {
             System.arraycopy(argTypes, end, newArgs, start, argTypes.length - end);
         }
-        return ofTrusted(returnType, newArgs);
+        return ofValidated(returnType, newArgs);
     }
 
     @Override
     public MethodTypeDesc insertParameterTypes(int pos, ClassDesc... paramTypes) {
         if (pos < 0 || pos > argTypes.length)
             throw new IndexOutOfBoundsException(pos);
 
+        for (ClassDesc param : paramTypes)
+            validateArgument(param);
```

Comment:
This check should be performed as part of copying `paramTypes` to `newArgs` to avoid TOC/TOU issues, e.g.:
```java
@Override
public MethodTypeDesc insertParameterTypes(int pos, ClassDesc... paramTypes) {
	if (pos < 0 || pos > argTypes.length)
		throw new IndexOutOfBoundsException(pos);

	ClassDesc[] newArgs = new ClassDesc[argTypes.length + paramTypes.length];
	if (pos > 0) {
		System.arraycopy(argTypes, 0, newArgs, 0, pos);
	}
	for (int i = 0; i < paramTypes.length; i++) {
		ClassDesc param = paramTypes[i];
		validateArgument(param);
		newArgs[pos + i] = param;
	}
	if (pos < argTypes.length) {
		System.arraycopy(argTypes, pos, newArgs, pos + paramTypes.length, argTypes.length - pos);
	}
	return ofValidated(returnType, newArgs);
}
```

See also:
https://github.com/openjdk/jdk/blob/230fac80f25e9608006c8928a8a7708bf13a818c/src/java.base/share/classes/java/util/ImmutableCollections.java#L186-L195

Reply:
Nice catch! I conservatively just move the validation loop after to keep the arraycopying.